src/nvim/screen.c

```diff
@@ -7027,6 +7033,28 @@ static void draw_tabline(void)
   redraw_tabline = FALSE;
 }
 
+// send tabline update to external ui
+void ui_ext_tabline_update(void)
+{
+  Array args = ARRAY_DICT_INIT;
+  ADD(args, INTEGER_OBJ(curtab->handle));
+  Array tabs = ARRAY_DICT_INIT;
+  FOR_ALL_TABS(tp) {
+    win_T *cwp = (tp == curtab) ? curwin : tp->tp_curwin;
+    get_trans_bufname(cwp->w_buffer);
+    Array tab = ARRAY_DICT_INIT;
+    ADD(tab, INTEGER_OBJ(tp->handle));
```

Comment:
It is awkward due to limitations in neovim-lua. In the python-client (and similarly in other plugin hosts) you get a useful object you can call methods on, assign to `nvim.current.tabpage` etc. The lua client should at least be fixed to decode the id so that it prints like `{[id]=1}`.